include/ginkgo/core/distributed/index_map.hpp

```diff
@@ -0,0 +1,176 @@
+struct index_map {
+    using partition_type = Partition<LocalIndexType, GlobalIndexType>;
+
+    /**
+     * \brief Maps global indices to local indices
+     *
+     * \param global_ids  the global indices to map
+     * \param index_space_v  the index space in which the returned local indices
+     *                       are defined
+     *
+     * \return  the mapped local indices. Any global index that is not in the
+     *          specified index space is mapped to invalid_index.
+     */
+    array<LocalIndexType> get_local(
+        const array<GlobalIndexType>& global_ids,
+        index_space index_space_v = index_space::combined) const;
+
+    /**
+     * \brief get size of index_space::local
+     */
+    size_type get_global_size() const;
+
+    /**
+     * \brief get size of index_space::local
+     */
+    size_type get_local_size() const;
+
+    /**
+     * \brief get size of index_space::non_local
+     */
+    size_type get_non_local_size() const;
+
+    /**
+     * \brief Creates a new index map.
+     *
+     * The passed in recv_connections may contain duplicates, which will be
+     * filtered out.
+     *
+     * \param exec  the executor
+     * \param partition  the partition of the global index set
+     * \param rank  the id of the global index space subset
+     * \param recv_connections  the global indices that are not owned by this
+     *                          rank, but accessed by it
+     */
+    index_map(std::shared_ptr<const Executor> exec,
+              std::shared_ptr<const partition_type> partition,
+              comm_index_type rank,
+              const array<GlobalIndexType>& recv_connections);
+
+    /**
+     * \brief Creates an empty index map.
+     */
+    index_map(std::shared_ptr<const Executor> exec);
+
+    /**
+     * \brief get the index set $R_k$ for this rank.
+     *
+     * The indices are ordered by their owning rank and global index.
+     */
+    const segmented_array<GlobalIndexType>& get_remote_global_idxs() const;
+
+    /**
+     * \brief get the index set $R_k$, but mapped to their respective local
+     *        index space.
+     *
+     * The indices are ordered by their owning rank and global index.
+     *
+     * The set $R_k = \hat{I}_k \setminus I_k$ can also be written as the union
+     * of the intersection of $\hat{I}_k$ with other disjoint sets
+     * $I_l, l \neq k$, i.e.
+     * $R_k = \bigcup_{j \neq k} \hat{I}_k \cap I_j = \bigcup_{j \neq k}
+     * R_{k,j}$. The set $R_{k,j}$ can then be mapped by $l_j$ to get the local
+     * indices wrt. part $j$. The indices here are mapped by $l_j$.
+     */
+    const segmented_array<LocalIndexType>& get_remote_local_idxs() const;
+
+    /**
+     * \brief get the rank ids which contain indices accessed by this rank.
+     *
+     * The order matches the order of the sets in get_remote_global_idxs and
+     * get_remote_local_idxs.
+     */
+    const array<comm_index_type>& get_remote_target_ids() const;
```

Comment:
I guess so, maybe I can actually do the same as for csr, i.e. storing the offsets, and local/global indices explicitly.

Reply:
`get_remote_target_ids` definitely makes sense, I'm just wondering whether `get_remote_global_idxs` will stay that way. But since this is still `experimental`, maybe we can replace it by a more efficient representation before stabilizing the interface.

Reply:
> I guess so, maybe I can actually do the same as for csr, i.e. storing the offsets, and local/global indices explicitly.

I am worried less about storage overhead than potential for inconsistencies between the two arrays, I'd prefer a single source of truth for the ranges.

Reply:
I think I will offer `get_remote_target_sizes` instead of the indices. This is more directly what I need, and then we could still change the internal representation.

Reply:
Does that map well to sparse communication? Otherwise we might have issues with scaling to huge numbers of ranks

Reply:
I realized that I actually use `get_remote_local_idxs` to communicate the local indices for the distributed row-gather. And they are also used without the row-gatherer in the `read_distributed`. So for this one I would at least need direct pointer access.

Reply:
I think I also provided `get_remote_global_idxs`, because this is directly defined through the class documentation, while the local indices need more explaining. So I could remove the global indices (but I feel a bit uncomfortable with that because it breaks some sort of symmetry).

Reply:
For now, I would leave it as it is. I specifically thought of segmented arrays as storage for these arrays, so changing that seems to have quite far-reaching consequences. Since this all still experimental, we could change it if we find better solutions later on.